clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java

```diff
@@ -110,78 +112,107 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
+                        remainingTimeToBlockNs -= timeNs;
+                        // check if we can satisfy this request from the free list,
+                        // otherwise allocate memory
+                        if (accumulated == 0 && size == this.poolableSize && !this.free.isEmpty()) {
+                            // just grab a buffer from the free list
+                            buffer = this.free.pollFirst();
+                            accumulated = size;
+                            restoreAvailableMemoryOnFailure = false;
+                        } else {
+                            // we'll need to allocate memory, but we may only get
+                            // part of what we need on this iteration
+                            freeUp(size - accumulated);
+                            int got = (int) Math.min(size - accumulated, this.availableMemory);
+                            this.availableMemory -= got;
+                            accumulated += got;
+                        }
                     }
 
-                    remainingTimeToBlockNs -= timeNs;
-                    // check if we can satisfy this request from the free list,
-                    // otherwise allocate memory
-                    if (accumulated == 0 && size == this.poolableSize && !this.free.isEmpty()) {
-                        // just grab a buffer from the free list
-                        buffer = this.free.pollFirst();
-                        accumulated = size;
-                    } else {
-                        // we'll need to allocate memory, but we may only get
-                        // part of what we need on this iteration
-                        freeUp(size - accumulated);
-                        int got = (int) Math.min(size - accumulated, this.availableMemory);
-                        this.availableMemory -= got;
-                        accumulated += got;
-                    }
-                }
-
-                // remove the condition for this thread to let the next thread
-                // in line start getting memory
-                Condition removed = this.waiters.removeFirst();
-                if (removed != moreMemory)
-                    throw new IllegalStateException("Wrong condition: this shouldn't happen.");
+                    // remove the condition for this thread to let the next thread
+                    // in line start getting memory
+                    Condition removed = this.waiters.removeFirst();
+                    if (removed != moreMemory)
+                        throw new IllegalStateException("Wrong condition: this shouldn't happen.");
 
+                    if (buffer == null)
+                        buffer = allocateByteBuffer(size);
+                    restoreAvailableMemoryOnFailure = false;
+                    //unlock happens in top-level, enclosing finally
+                    return buffer;
+                } finally {
```

Comment:
Not without additional complications.  One finally deals with cases caused by waiting for memory to become available the other deals with cases causes by just allocating memory.